File: bigcommerce-mcp-server/src/parsers/schema_parser.py
"""Parser for JSON schema files"""

import json
import yaml
import logging
from pathlib import Path
from typing import Dict, List, Any, Optional

logger = logging.getLogger(__name__)


class SchemaParser:
    """Parser for JSON schema files"""
    
    def __init__(self, models_path: Path):
        self.models_path = models_path
        self.schemas = {}
    
    async def load_all_schemas(self) -> Dict[str, Any]:
        """Load all JSON schema files"""
        schemas = {}
        
        if not self.models_path.exists():
            logger.warning(f"Models path does not exist: {self.models_path}")
            return schemas
        
        # Find all JSON and YAML schema files
        schema_files = (
            list(self.models_path.glob("**/*.json")) + 
            list(self.models_path.glob("**/*.yml")) + 
            list(self.models_path.glob("**/*.yaml"))
        )
        
        for schema_file in schema_files:
            try:
                schema_data = await self._load_schema_file(schema_file)
                if schema_data:
                    schema_name = self._extract_schema_name(schema_file)
                    schemas[schema_name] = {
                        'file_path': str(schema_file),
                        'schema': schema_data,
                        'name': schema_name
                    }
                    logger.info(f"Loaded schema: {schema_name}")
                
            except Exception as e:
                logger.error(f"Error loading schema file {schema_file}: {e}")
        
        self.schemas = schemas
        return schemas
    
    async def _load_schema_file(self, file_path: Path) -> Optional[Dict[str, Any]]:
        """Load a single schema file"""
        try:
            with open(file_path, 'r', encoding='utf-8') as file:
                if file_path.suffix.lower() == '.json':
                    return json.load(file)
                else:
                    return yaml.safe_load(file)
        except Exception as e:
            logger.error(f"Error reading schema file {file_path}: {e}")
            return None
    
    def _extract_schema_name(self, file_path: Path) -> str:
        """Extract schema name from file path"""
        # Remove file extension and use relative path
        relative_path = file_path.relative_to(self.models_path)
        return str(relative_path.with_suffix(''))
    
    def get_schema(self, schema_name: str) -> Optional[Dict[str, Any]]:
        """Get a specific schema"""
        return self.schemas.get(schema_name)
    
    def search_schemas(self, query: str, limit: int = 10) -> List[Dict[str, Any]]:
        """Search schemas by name or content"""
        results = []
        query_lower = query.lower()
        
        for schema_name, schema_data in self.schemas.items():
            score = 0
            
            # Check schema name
            if query_lower in schema_name.lower():
                score += 10
            
            # Check schema content (basic string search)
            schema_str = json.dumps(schema_data['schema'], default=str).lower()
            if query_lower in schema_str:
                score += 3
            
            if score > 0:
                results.append({
                    'name': schema_name,
                    'score': score,
                    'data': schema_data
                })
        
        # Sort by score and limit results
        results.sort(key=lambda x: x['score'], reverse=True)
        return results[:limit] 

File: src/parsers/schema-parser.ts
import { readFile, readdir, stat } from 'node:fs/promises';
import { join, extname, relative, basename } from 'node:path';
import { parse } from 'yaml';
import { SchemaFile } from '../types/index.js';

export class SchemaParser {
  private schemas: Map<string, SchemaFile> = new Map();
  private modelsPath: string;

  constructor(modelsPath: string) {
    this.modelsPath = modelsPath;
  }

  async loadAllSchemas(): Promise<Map<string, SchemaFile>> {
    try {
      const files = await this.getSchemaFiles();
      
      for (const file of files) {
        try {
          const schema = await this.loadSchema(file);
          if (schema) {
            this.schemas.set(schema.name, schema);
          }
        } catch (error) {
          console.warn(`Failed to load schema ${file}:`, error);
        }
      }

      console.log(`Loaded ${this.schemas.size} schema files`);
      return this.schemas;
    } catch (error) {
      console.error('Error loading schemas:', error);
      return new Map();
    }
  }

  private async getSchemaFiles(): Promise<string[]> {
    const files: string[] = [];
    
    const scanDirectory = async (dir: string): Promise<void> => {
      try {
        const items = await readdir(dir);
        
        for (const item of items) {
          const fullPath = join(dir, item);
          const stats = await stat(fullPath);
          
          if (stats.isDirectory()) {
            await scanDirectory(fullPath);
          } else if (stats.isFile()) {
            const ext = extname(item).toLowerCase();
            if (ext === '.yaml' || ext === '.yml' || ext === '.json') {
              files.push(fullPath);
            }
          }
        }
      } catch (error) {
        console.warn(`Error scanning directory ${dir}:`, error);
      }
    };

    await scanDirectory(this.modelsPath);
    return files;
  }

  private async loadSchema(filePath: string): Promise<SchemaFile | null> {
    try {
      const content = await readFile(filePath, 'utf-8');
      const ext = extname(filePath).toLowerCase();
      
      let schema: any;
      if (ext === '.json') {
        schema = JSON.parse(content);
      } else {
        schema = parse(content);
      }

      const name = basename(filePath, extname(filePath));

      return {
        path: relative(this.modelsPath, filePath),
        name,
        schema,
      };
    } catch (error) {
      console.error(`Error parsing schema ${filePath}:`, error);
      return null;
    }
  }

  getSchema(schemaName: string): SchemaFile | null {
    return this.schemas.get(schemaName) || null;
  }

  searchSchemas(query: string): SchemaFile[] {
    const results: Array<{ schema: SchemaFile; score: number }> = [];
    const queryLower = query.toLowerCase();

    for (const schema of this.schemas.values()) {
      let score = 0;

      // Name match
      if (schema.name.toLowerCase().includes(queryLower)) {
        score += 10;
      }

      // Schema content match
      const schemaText = JSON.stringify(schema.schema).toLowerCase();
      const matches = (schemaText.match(new RegExp(queryLower, 'g')) || []).length;
      score += matches;

      if (score > 0) {
        results.push({ schema, score });
      }
    }

    return results
      .sort((a, b) => b.score - a.score)
      .slice(0, 20)
      .map(r => r.schema);
  }

  getAllSchemas(): Map<string, SchemaFile> {
    return this.schemas;
  }
} 